cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -187,31 +188,32 @@ struct ProductImpl : public ScalarAggregator {
     const auto& other = checked_cast<const ThisType&>(src);
     this->count += other.count;
     this->product =
-        static_cast<ProductType>(to_unsigned(this->product) * to_unsigned(other.product));
+        MultiplyTraits<AccType>::Multiply(*out_type, this->product, other.product);
     return Status::OK();
   }
 
   Status Finalize(KernelContext*, Datum* out) override {
     if (this->count < options.min_count) {
-      out->value = std::make_shared<OutputType>();
+      out->value = std::make_shared<OutputType>(out_type);
     } else {
-      out->value = MakeScalar(this->product);
+      out->value = std::make_shared<OutputType>(this->product, out_type);
     }
     return Status::OK();
   }
 
-  size_t count = 0;
-  typename AccType::c_type product = 1;
+  std::shared_ptr<DataType> out_type;
```

Comment:
Nit: const reference?

Reply:
I don't think we want this as the reference may be invalidated.